src/games/seabattle/ai.ts

```diff
@@ -0,0 +1,191 @@
+        if (otherHitSalvos.length >= 2) {
+          // we have hit this ship at least twice, so we can make an intelligent move
+          return this.getMoveFromOtherSalvos(state, otherHitSalvos);
+        }
+        // no other salvos for the same ship were hit, so hit a random neighbor
+        return this.getRandomNeighbor(state, salvo);
+      }
+    }
+    // generate a random move
+    return this.generateRandomMove(state);
+  }
+
+  getMoveFromOtherSalvos(state: IPlayState, hitSalvos: ISalvo[]): ICell {
+    const len = hitSalvos.length;
+    const lastHitSalvo = hitSalvos[len - 1];
+    const lastHitCell = lastHitSalvo.cell;
+    const secondLastHitCell = hitSalvos[len - 2].cell;  // second to last hit
+    const firstHitCell = hitSalvos[0].cell;
+
+    let move: ICell = { x: lastHitCell.x, y: lastHitCell.y };  // set to last hit cell
+    let moveDelta;
+    if (lastHitCell.x === secondLastHitCell.x) {
+      // ship is vertical x, so move horizontal y
+      moveDelta = lastHitCell.y - secondLastHitCell.y;
+      if (moveDelta !== -1 && moveDelta !== 1) {
+        return this.getRandomNeighbor(state, lastHitSalvo);
+      }
+      if (this.isInBounds(move.y + moveDelta)) {
+        move = { x: move.x, y: move.y + moveDelta };  // reset to first hit cell
+      } else {
+        move = { x: firstHitCell.x, y: firstHitCell.y - moveDelta };  // reset to first hit cell
+      }
+    } else if (lastHitCell.y === secondLastHitCell.y) {
+      // ship is horizontal y, so move vertical x
+      moveDelta = lastHitCell.x - secondLastHitCell.x;
+      if (moveDelta !== -1 && moveDelta !== 1) {
+        return this.getRandomNeighbor(state, lastHitSalvo);
+      }
+      if (this.isInBounds(move.x + moveDelta)) {
+        move = { x: move.x + moveDelta, y: move.y };  // reset to first hit cell
+      } else {
+        move = { x: firstHitCell.x - moveDelta, y: firstHitCell.y };  // reset to first hit cell
+      }
+    } else {
+      // we hit a different ship than the one before, so hit a random neighbor:
+      return this.getRandomNeighbor(state, lastHitSalvo);
+    }
+    return move;
+  }
+
+  isInBounds(x: number) {
+    const valid = (x >= 0 && x <= 9);
+    return valid;
+  }
+
+  isValidMove(state: IPlayState, cell: ICell) {
+    const valid = (this.isInBounds(cell.x) && this.isInBounds(cell.y) && this.isUniqueMove(state, cell));
+    return valid;
+  }
+
+  areVerticalMovesPossible(state: IPlayState, move: ICell) {
+    const possible = (this.isValidMove(state, move));
+  }
+
+  getRandomNeighbor(state: IPlayState, salvo: ISalvo): ICell {
+    let i = 0;
+    let vertical;
+    let validMove;
+    let randomChange;
+    let move: ICell;
+    while (!validMove && i < 10) {
+      vertical = this.genRandomNumber(0, 1) === 0;
+      move = { x: salvo.cell.x, y: salvo.cell.y };
+      randomChange = (this.genRandomNumber(0, 1) === 0) ? -1 : 1;
+      if (vertical) {
```

Comment:
this would look simpler:
moveDelta = { x: (vertical) ? randomChange : 0, y: (vertical) ? 0 : randomChange };
move = { x: salvo.cell.x + moveDelta, y: salvo.cell.y + moveDelta }
Given that you dont need to check for the bounds (it is done in isValidMove), this should be enough

Reply:
Thanks!